docs: add debouncing section to signal forms async operations (#68334)
angular/angular: adev/src/content/guide/forms/signals/async-operations.md

### Custom commit timing

For commit logic that a duration or `'blur'` can't express, pass a `Debouncer` function. The function receives the field context and an `AbortSignal`, and returns a `Promise<void>` that resolves when the model should update:

```ts
import {debounce, type Debouncer} from '@angular/forms/signals';

const shorterWhenLonger: Debouncer<string> = ({value}, abortSignal) => {
  // Shorter queries get a longer delay since the user is likely still typing.
  const ms = value().length < 3 ? 500 : 200;
  return new Promise((resolve) => {
    const timeoutId = setTimeout(resolve, ms);
    // Abort fires when this field is touched or its value changes, so the pending timer is cleared
    abortSignal.addEventListener(
      'abort',
      () => {
        clearTimeout(timeoutId);
        resolve();
      },
      {once: true},
    );
  });
};

form(this.registrationModel, (schemaPath) => {
  debounce(schemaPath.username, shorterWhenLonger);
});
```

> The `abortSignal` fires when the field is touched, or when its value changes before the debounce resolves. Resolve the promise on abort so your debouncer releases any pending timers. The framework commits the pending value on touch, and discards it when a newer value arrives. See the [`debounce` API reference](api/forms/signals/debounce) for the full `Debouncer` signature.

Reviewer: wdyt adding a link to the [AbortSignal MDN](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) docs for `abortSignal`?

Author: Great suggestion!

The `rxResource` function works directly with Observables and handles subscription cleanup automatically when the field value changes.

## Debouncing

> The `debounce` rule delays when a user's input is committed to the form model. You can think of it as debounce holding back the values until the user pauses typing. This is useful when downstream behavior shouldn't react to every keystroke, such as expensive derived computations, validation that flashes errors mid-word, or search filters that reapply on each character.

Reviewer 2: Should we mention somewhere that `controlValue` is updated instantly but `value` is only updated after the debounce time ?

Author: Definitely. I've added it to the "Touch flushes the model" h3 since it's more of an advanced concept and this is still an introductory part of this section.

      <span class="checking">Checking availability...</span>
    }
  `,
})
export class Registration {
  registrationModel = signal({username: ''});

  registrationForm = form(this.registrationModel, (schemaPath) => {
    // Hold UI updates for 300 ms before writing to the model
    debounce(schemaPath.username, 300);

    // Runs against the debounced model value, not every keystroke
    validateHttp(schemaPath.username, {
      request: ({value}) => {
        const username = value();
        // Skip the request for blank values
        return username ? `/api/users/check?username=${username}` : undefined;
      },
      onSuccess: (response) =>
        response.available ? null : {kind: 'usernameTaken', message: 'Username is already taken'},
      onError: () => ({
        kind: 'serverError',
        message: 'Could not verify username availability',
      }),
    });
  });
}
```

With a 300 ms debounce, the model updates and validates only after the user pauses typing longer than the configured duration. For example, typing "signal forms" in a quick burst fires one validation request instead of twelve.

### Touch flushes the model

> Regardless of the debounce duration, a field's value writes to the model immediately when the field becomes touched. Native inputs become touched on blur, so a user who finishes typing and tabs away doesn't have to wait for the debounce timer to expire. Custom controls can mark the field as touched in response to any event they choose.

Reviewer 2: wdyt of this: 
```suggestion
Regardless of the debounce duration,  the framework writes the field's value to the model immediately when the field becomes touched. Native inputs become touched on blur, so a user who finishes typing and tabs away doesn't have to wait for the debounce timer to expire. Custom controls can mark the field as touched in response to any event they choose.
```

Maybe we _could_ replace "the field's value " with "the fields `controlValue` to the model"

This matters for form submission in the typical case. When the user clicks a submit button, the focused input blurs, which touches that field and flushes its pending debounce before the submission handler runs.

### Commit only on blur

> Some fields shouldn't update mid-typing at all, and instead should only update after the user has finished entering a value. For example, instead of a search filter that reapplies on every change, or a form that triggers expensive derived state, it makes more sense for the model to wait until the user is done.

Reviewer 2: We could give a more active tone here: 

```suggestion
Some fields shouldn't update mid-typing at all, and instead should only update after the user has finished entering a value. For example, if you have a search filter that reapplies on every change or a form that triggers expensive derived state, it is often better for the model to wait until the user finishes typing.
```

Choose between the two layers based on scope:

| Option                                                        | When to use                                                                                                                         |
| ------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------- |
| `debounce()` rule                                             | Sync validation, derived state, and submission should all wait until the field commits. The whole field shouldn't react mid-typing. |
| `validateHttp({ debounce })` or `validateAsync({ debounce })` | Cheap sync validators should give immediate feedback, but expensive async calls should wait for the user to pause.                  |

Both options accept a duration in milliseconds. Their custom-timing callbacks differ: the form-level rule takes a `Debouncer`, and the validator-level option takes a `DebounceTimer` from `@angular/core`. The two signatures are not interchangeable.

## Composing resources in async validation with a factory

> The built-in `debounce` option covers throttling, but `validateAsync()` exposes a deeper composition point: the `factory` function. The factory receives the params as a signal and returns a resource, otherwise you are free to compose whatever you need between those two points.

Reviewer 2: Q: Should we add `debounce` to `LINK_EXEMPT` in `adev/shared-docs/pipeline/shared/linking.mts` ? 

"`debounce` option" links to debounce standalone function. 
The alternative would be that we link `debounce` ourselves to `validateAsync`.

Author: Ahhh good question. I don't want to impact the linking of other guides to the actual debounce function, so I'll update it to be manually linked.

> In its simplest form, a factory wraps a single resource. A username-availability check can live as a named class-field method on the component, then get wired into `validateAsync` by reference:

Reviewer 2: ```suggestion
In its simplest form, a factory wraps a single resource. A username-availability check can live as a method on the component class, and then get wired into validateAsync by reference:":
```
